SNP bit checks on the host and ubuntu guest from instruction set

LakshmiSaiHarika opened the pull request:
1. Added CPU check on the host for all host SEV features from cpuid 0x8000001f 
2. Added host SNP enablement check in BIOS from MSR 0xC0010010 bit 23 & 24
3. Added additional check on the guest to see if guest security bits are active from  MSR 0xc0010131 (MSR_AMD64_SEV)

I validated all the added checks in this PR  as follows:

-   Tested CPU support check for SNP on Rome, milan to check the CPU support check which works fine
-   Tested SNP enablement check in Bios which shows correct error message when SME/SNP are disabled in BIOS
-   Tested guest SEV/SEV-ES/SNP bit status on SNP guest and normal guest(w/o SNP enabled)

larrydewey reviewed (changes requested):
Will you please run these through [shellcheck](https://shellcheck.net)? There are a number of changes which I would like to see applied for this PR before we approve it.

LakshmiSaiHarika commented:
Hi @larrydewey, I made few changes and did shellcheck... Please let me know if any further changes are required.

larrydewey commented on tools/snp.sh:
```diff
@@ -110,10 +110,12 @@ SEV_SNP_MEASURE_VERSION="0.0.11"
+  >&2 echo "    launch-guest             Launch a SNP guest"
+  >&2 echo "    attest-guest             Use virtee/snpguest and sev-snp-measure to attest a SNP guest"
+  >&2 echo "    stop-guests              Stop all SNP guests started by this script"
+  >&2 echo "    check-snp-on-host-msr    Perform SNP Check on the host via msr check"
+  >&2 echo "    check-snp-on-guest-msr    Perform SNP Check on the guest via msr check"
```
```suggestion
  >&2 echo "    check-host-snp-msr    Read the dedicated host MSR to determine if SNP is on and supported"
  >&2 echo "    check-guest-snp-msr   Read the dedicated guest MSR to determine if SNP is on and supported"
```

larrydewey commented on tools/snp.sh:
```diff
@@ -148,6 +150,12 @@ cleanup() {
       stop-guests)
         ;;
 
+      check-snp-on-host-msr)
+        ;;
+
+      check-snp-on-guest-msr)
+        ;;
```
```suggestion
      check-host-snp-msr)
        ;;

      check-guest-snp-msr)
        ;;
```

larrydewey commented on tools/snp.sh:
```diff
@@ -164,6 +172,63 @@ verify_snp_host() {
+  local host_cpuid_eax
+  host_cpuid_eax=$(cpuid -1 -r -l 0x8000001f | grep -oE "eax=[[:alnum:]]+ " | cut -c5- | tr -d '[:space:]')
+
+  # Bit 0 indicates support for SME status
+  local sme_bit=1
+  local sme_status=$((host_cpuid_eax & sme_bit))
+
+  #  Bit 1 indicates support for SEV 
+  local sev_bit=2
+  local sev_status=$((host_cpuid_eax & sev_bit))
+
+  # Bit 3 indicates support for SEV-ES
+  local sev_es_bit=8
+  local sev_es_status=$((host_cpuid_eax & sev_es_bit))
+
+  # Bit 4 indicates support for SNP
+  local snp_bit=16
+  local snp_status=$((host_cpuid_eax & snp_bit))
+
+  # Map all sev features in a single associative array for all SEV features support in the processor
+  declare -A all_actual_results=(
+    [SME]=${sme_status} 
+    [SEV]=${sev_status}
+    [SEV-ES]=${sev_es_status} 
+    [SNP]=${snp_status}
+  )
+
+  declare -A results_to_match=(
+    [SME]=1 
+    [SEV]=2
+    [SEV-ES]=8
+    [SNP]=16
+    )
```
Is there a reason we are declaring this twice?

LakshmiSaiHarika replied:
I have done code cleanup for this

larrydewey commented on tools/snp.sh:
```diff
@@ -848,6 +913,68 @@ verify_snp_guest() {
+  ssh_guest_command "sudo DEBIAN_FRONTEND=noninteractive sudo apt install -y msr-tools > /dev/null 2>&1"
+  ssh_guest_command "sudo modprobe msr"
+
+  # Get the guest (MSR_AMD64_SEV) value
+  local guest_msr_read
+  guest_msr_read=$(ssh_guest_command "sudo rdmsr -p 0 0xc0010131")
+  guest_msr_read=$(echo "$guest_msr_read" | tr -d '\r' | bc)
+
+  # Bit 0 indicates support for Guest SEV feature 
+  local sev_bit=1
+  sev_status=$((guest_msr_read & sev_bit))
+
+ 
+  # Bit 1 indicates support for Guest SEV-ES feature 
+  local sev_es_bit=2
+  sev_es_status=$((guest_msr_read & sev_es_bit))
+
+
+  # Bit 2 indicates support for Guest SEV-SNP feature 
+  local snp_bit=4
+  snp_status=$((guest_msr_read & snp_bit))
+
+  # Map all sev features in a single associative array for all guest SEV features
+  declare -A all_actual_results=(
+    [SEV]=${sev_status}
+    [SEV-ES]=${sev_es_status} 
+    [SNP]=${snp_status}
+  )
+
+  declare -A results_to_match=( 
+    [SEV]=1
+    [SEV-ES]=2
+    [SNP]=4
+  )
```
We are declaring it twice for a second time?...

larrydewey commented on tools/snp.sh:
```diff
@@ -1086,6 +1213,16 @@ main() {
         shift
         ;;
 
+      check-snp-on-host-msr)
+        COMMAND="check-snp-on-host-msr"
+        shift
+        ;;
+
+      check-snp-on-guest-msr)
+        COMMAND="check-snp-on-guest-msr"
+        shift
+        ;;
```
```suggestion

      check-host-snp-msr)
        COMMAND="check-host-snp-msr"
        shift
        ;;

      check-guest-snp-msr)
        COMMAND="check-guest-snp-msr"
        shift
        ;;
```

larrydewey replied:
Rename host from msr to cpuid

larrydewey commented on tools/snp.sh:
```diff
@@ -1113,6 +1250,7 @@ main() {
       ;;
 
     setup-host)
+      verify_if_host_is_snp_capable
```
Wouldn't it make sense to have a single function for this and pass a parameter to determine guest/host?

LakshmiSaiHarika replied:
As per this, I created a single function "verify_platform_snp_bit_status" with guest/host flag to handle snp bit check

larrydewey commented on tools/snp.sh:
```diff
@@ -1164,6 +1304,14 @@ main() {
       stop_guests
       ;;
 
+    check-snp-on-host-msr)
+      verify_if_host_is_snp_capable
+      ;;
+
+    check-snp-on-guest-msr)
+      verify_guest_snp_bit_status_from_msr
+      ;;
+
```
See comments above

larrydewey commented on tools/snp.sh:
```diff
@@ -164,6 +172,63 @@ verify_snp_host() {
   fi
 }
 
+verify_if_host_is_snp_capable() {
```
Would it make sense to just create a single function here and pass parameters to determine guest/host?

LakshmiSaiHarika replied:
As per this, I created a single function "verify_platform_snp_bit_status" with guest/host parameter to handle snp bit check

LakshmiSaiHarika commented:
Hi @larrydewey I addressed the changes as per the conversation above, please let me know if any additional changes are required

DGonzalezVillal commented on docs/snp.md:
````diff
@@ -52,6 +52,12 @@ wget https://github.com/amd/sev-utils/raw/main/tools/snp.sh
+./snp.sh check-host-snp-cpuid
+```
+
+
````
Do we want this to be it's own individual command? I see that it is being called when people are setting up the host and launching the guest. Is there additional functionality to be it's own individual command?

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -164,6 +172,83 @@ verify_snp_host() {
+    # Get the host cpuid eax
+    local host_cpuid_eax
+    host_cpuid_eax=$(cpuid -1 -r -l 0x8000001f | grep -oE "eax=[[:alnum:]]+ " | cut -c5- | tr -d '[:space:]')
+
```
I had created a function that does this for you called get_cpuid if you want to use that instead.
https://github.com/amd/sev-utils/blob/4419036ac8790614f429134b16a7427809b54170/tools/snp.sh#L933

LakshmiSaiHarika replied:
done

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -164,6 +172,83 @@ verify_snp_host() {
+    host_cpuid_eax=$(cpuid -1 -r -l 0x8000001f | grep -oE "eax=[[:alnum:]]+ " | cut -c5- | tr -d '[:space:]')
+
+    # Map all the computed host sev/snp bit values in a single associative array
+    declare -A actual_sev_snp_bit_status=(
+      [SME]=$((${host_cpuid_eax} & 1))
+      [SEV]=$((${host_cpuid_eax} & 2))
+      [SEV-ES]=$((${host_cpuid_eax} & 8))
+      [SNP]=$((${host_cpuid_eax} & 16))
+    )
+
+    # Map all the expected host sev/snp bit values in a single associative array
+    declare -A expected_sev_snp_bit_value=(
+      [SME]=1
+      [SEV]=2
+      [SEV-ES]=8
+      [SNP]=16
+      )
```
You could just grab the specific bit for each of the features and that way you don't have to create two different arrays.

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -164,6 +172,83 @@ verify_snp_host() {
+    guest_msr_read=$(echo "${guest_msr_read}" | tr -d '\r' | bc)
+
+    # Map all the sev features in a single associative array for all guest SEV features
+    declare -A actual_sev_snp_bit_status=(
+      [SEV]=$((${guest_msr_read} & 1))
+      [SEV-ES]=$((${guest_msr_read} & 2))
+      [SNP]=$((${guest_msr_read} & 4))
+    )
+
+    declare -A expected_sev_snp_bit_value=(
+      [SEV]=1
+      [SEV-ES]=2
+      [SNP]=4
+    )
```
Same thing as above. You can just get the specific bit for each feature and that way you don't need 2 different arrays.

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -1228,6 +1325,7 @@ main() {
 
       setup_and_launch_guest
       wait_and_retry_command verify_snp_guest
+      wait_and_retry_command verify_platform_snp_bit_status guest
```
This should maybe replace verify_snp_guest. They are essentially doing the same thing, the only difference is that the original function is checking dmesg, while the bit_status is looking at the guest MSR to verify enablement. No need to run both of them if the MSR check is working correctly.

LakshmiSaiHarika replied:
Ok, I commented verify_snp_guest above

DGonzalezVillal reviewed (commented):
Just some nits and some overall functionality questions. But overall it lgtm.

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -164,6 +163,72 @@ verify_snp_host() {
+      [SEV-ES]=$(( (${host_cpuid_eax} >> 3) & 1))
+      [SNP]=$(( (${host_cpuid_eax} >> 4) & 1))
+    )
+
```
Additionally to the CPUID (which checks if the feature is supported by the CPU) there is also an MSR that you can use to check if SME and SNP are enabled.

MSR 0xC0010010 bit 23 checks SME
MSR 0xC0010010 bit 24 checks SNP

LakshmiSaiHarika replied:
Added verify_host_snp_enablement module to perform the enablement check

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -164,6 +163,72 @@ verify_snp_host() {
+      return 1
+    fi
+
+    # Install guest rdmsr package dependencies to insert & insert guest msr module
+    ssh_guest_command "sudo DEBIAN_FRONTEND=noninteractive sudo apt install -y msr-tools > /dev/null 2>&1" > /dev/null 2>&1
+    ssh_guest_command "sudo modprobe msr" > /dev/null 2>&1
```
Will this only work on Debian?

LakshmiSaiHarika replied:
Oh.. good catch, yes this works only on debian, I will make sure to add these under the OS condition

LakshmiSaiHarika replied:
> Will this only work on Debian?

For now yes, can easily add additional guest linux os support in this modififed PR

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -164,6 +163,72 @@ verify_snp_host() {
+
+    # Read the guest (MSR_AMD64_SEV) value
+    local guest_msr_read
+    guest_msr_read=$(ssh_guest_command "sudo rdmsr -p 0 0xc0010131")
+    guest_msr_read=$(echo "${guest_msr_read}" | tr -d '\r' | bc)
```
In one of my tools that I built out to do this, I had an issue with  padding. If I had a bit number that looked like 0000001, my msr read would return 1 not the full number. So when I tried to reach a bit that was a 0 I'd get a an issue trying to reach a digit that was not there. Have you checked for this issue? Try running this command in a guest with no SEV enabled and see if the fail case doesn't cause an error.

LakshmiSaiHarika replied:
I verified guest_msr_read for the guest with no sev feature, I could exit the guest launch with error code 1 and error message highlighted in the verify_platform_status()

LakshmiSaiHarika committed:
snp.sh: Verify if SNP(security) bits are set

This module verifies if all the security bits are set to 1 for any given instruction set

Signed-off-by: Harika Nittala <lnittala@amd.com>

LakshmiSaiHarika committed:
snp.sh: Added CPUID check to verify the SNP feature support in host CPU

This verifies if CPU is capable of SNP based on the SNP bit value present in the CPUID 0x8000001f

Signed-off-by: Harika Nittala <lnittala@amd.com>

LakshmiSaiHarika committed:
snp.sh: Added MSR check to confirm if SNP is enabled in host BIOS

This verifies if SME, SNP are enabled in the host BIOS settings by reading SME and SNP bit status from MSR 0xC0010010

Bit #23 corresponds to the SME bit status
Bit #24 corresponds to the SNP bit status

Signed-off-by: Harika Nittala <lnittala@amd.com>

LakshmiSaiHarika committed:
snp.sh: Added guest MSR check to verify SNP bit status

Added  MSR 0xC0010010 check to validate if guest SEV, SEV-ES and SNP are enabled by reading SEV, SEV-ES and SNP bits from MSR 0xC0010010 instruction set

Bit #0 corresponds to the SEV bit status
Bit #1 corresponds to SEV-ES bit status
Bit #2 corresponds to SNP bit status

Signed-off-by: Harika Nittala <lnittala@amd.com>

AdithyaKrishnan commented on tools/snp.sh:
```diff
@@ -156,6 +156,61 @@ cleanup() {
+verify_host_snp_enablement() {
+  echo -e "Verifying if SME, SNP are enabled in the host from MSR 0xC0010010..."
+
+  sudo modprobe msr
```
`if ! sudo modprobe msr; then`
`    >&2 echo "ERROR: Failed to load MSR kernel module. Ensure you have the necessary sudo permissions."`
`    return 1`
`fi`

AdithyaKrishnan commented on tools/snp.sh:
```diff
@@ -1099,6 +1154,55 @@ setup_guest_attestation() {
+
+  case ${guest_linux_distro} in
+    ubuntu)
+      ssh_guest_command "sudo DEBIAN_FRONTEND=noninteractive sudo apt install -y msr-tools > /dev/null 2>&1" > /dev/null 2>&1
```
Are we assuming that the user will have sudo permissions by default? 
If yes, disregard my comments.
Else, we might have to modify the code to handle that permission errors aren't misconstrued as actual errors

LakshmiSaiHarika replied:
yes, I tested this, this works by default

AdithyaKrishnan reviewed (commented):
I believe this code has already been reviewed multiple times. 
I couldn't find any major new suggestions apart from what was already provided to you.

DGonzalezVillal reviewed (approved):
lgtm
